### ISSUE TITLE
Missing type narrowing of self

### ISSUE THREAD

**Issue by jacobbogdanov** [microsoft/pyright#773]
**Describe the bug**
Pyright does not correctly narrow the type of `self` based on `isinstance` guard clause.

**To Reproduce**
See code below.

**Expected behavior**
adding `isinstance(self, X)` should narrow the type of `self` to only be `X` within that scope.
 
**Screenshots or Code**
```py
class Base:
    def get_value(self) -> int:
        if isinstance(self, Derived):
            return self.calculate()  # Cannot access member "calculate" for type "Base"
                                     # Member "calculate" is unknown
        return 7

class Derived(Base):
    def calculate(self) -> int:
        return 2 * 2
```

Interestingly enough, adding almost any type annotation to `self` in the function signature causes this error to go away. so
- `def get_value(self: Base) -> int:`
- `def get_value(self: Union[Base, "Derived"]) -> int:`
- `def get_value(self: Any) -> int:`
all resolve the issue.

**VS Code extension or command-line**
pyright CLI and vscode extension version `1.1.47`

**Additional context**

**Comment by erictraut** [comment]
This is technically a bug in Pyright, and I'll fix it, but you should be ashamed of yourself for this code. :) Base classes should _never_ have knowledge of classes that derive from them! That's a fundamental rule in object-oriented programming.

This will be fixed in the next release of Pyright.

**Comment by erictraut** [comment]
This is now fixed in Pyright version 1.1.48, which I just published. It will also be included in the next public release of Pylance.
